third_party/move/move-compiler-v2/src/file_format_generator/module_generator.rs

```diff
@@ -0,0 +1,599 @@
+            name_to_idx: Default::default(),
+            address_to_idx: Default::default(),
+            fun_to_idx: Default::default(),
+            struct_to_idx: Default::default(),
+            struct_def_inst_to_idx: Default::default(),
+            field_to_idx: Default::default(),
+            field_inst_to_idx: Default::default(),
+            types_to_signature: Default::default(),
+            cons_to_idx: Default::default(),
+            fun_inst_to_idx: Default::default(),
+            module,
+        };
+        gen.gen_module(ctx, module_env);
+        gen.module
+    }
+
+    /// Generates a module, visiting all of it's members.
+    fn gen_module(&mut self, ctx: &ModuleContext, module_env: &ModuleEnv<'_>) {
+        // Create the self module handle, at well known handle index 0
+        let loc = &module_env.get_loc();
+        self.module_index(ctx, loc, module_env.get_id());
+
+        for struct_env in module_env.get_structs() {
+            self.gen_struct(ctx, &struct_env)
+        }
+        for fun_env in module_env.get_functions() {
+            FunctionGenerator::run(self, ctx, fun_env);
+        }
+    }
+
+    /// Generate information for a struct.
+    fn gen_struct(&mut self, ctx: &ModuleContext, struct_env: &StructEnv<'_>) {
+        let loc = &struct_env.get_loc();
+        let struct_handle = self.struct_index(ctx, loc, struct_env);
+        let field_information = FF::StructFieldInformation::Declared(
+            struct_env
+                .get_fields()
+                .map(|f| {
+                    let name = self.name_index(ctx, loc, f.get_name());
+                    let signature =
+                        FF::TypeSignature(self.signature_token(ctx, loc, &f.get_type()));
+                    FF::FieldDefinition { name, signature }
+                })
+                .collect(),
+        );
+        let def = FF::StructDefinition {
+            struct_handle,
+            field_information,
+        };
+        ctx.checked_bound(
+            loc,
+            self.module.struct_defs.len(),
+            MAX_STRUCT_DEF_COUNT,
+            "struct",
+        );
+        self.module.struct_defs.push(def);
+    }
+
+    /// Obtains or creates an index for a signature, a sequence of types.
+    pub fn signature(
+        &mut self,
+        ctx: &ModuleContext,
+        loc: &Loc,
+        tys: Vec<Type>,
+    ) -> FF::SignatureIndex {
+        if let Some(idx) = self.types_to_signature.get(&tys) {
+            return *idx;
+        }
+        let tokens = tys
+            .iter()
+            .map(|ty| self.signature_token(ctx, loc, ty))
+            .collect::<Vec<_>>();
+        let idx = FF::SignatureIndex(ctx.checked_bound(
+            loc,
+            self.module.signatures.len(),
+            MAX_SIGNATURE_COUNT,
+            "signature",
+        ));
+        self.module.signatures.push(FF::Signature(tokens));
+        self.types_to_signature.entry(tys).or_insert(idx);
```

Comment:
If you're expecting concurrent modifications, then self.types_to_signature.entry(tys) should be empty here, so you should just do:
```
self.types_to_signature.insert(tys, idx);
```
OTOH, if some of the intervening code could modify self.types_to_signature.entry(tys), then shouldn't we be returning the same value here, rather than idx on line 165?

Reply:
This indeed should have been an `insert` and not an `entry`, just a cut and paste error. No concurrency is expected here, and AFAICT Rust wouldn't allow it anyway since some central data types (like `GlobEnv`) are not thread safe.